packages/cli/src/util/get-latest-version/index.ts

```diff
@@ -0,0 +1,151 @@
+import semver from 'semver';
+import XDGAppPaths from 'xdg-app-paths';
+import { dirname, parse as parsePath, resolve as resolvePath } from 'path';
+import type { Output } from '../output';
+import { existsSync, outputJSONSync, readJSONSync } from 'fs-extra';
+import type { PackageJson } from '@vercel/build-utils';
+import { spawn } from 'child_process';
+
+interface GetLatestVersionOptions {
+  cacheDir?: string;
+  distTag?: string;
+  output?: Output;
+  pkg: PackageJson;
+  updateCheckInterval?: number;
+}
+
+interface PackageInfoCache {
+  version: string;
+  expireAt: number;
+  notified: boolean;
+}
+
+interface GetLatestWorkerPayload {
+  cacheFile?: string;
+  distTag?: string;
+  updateCheckInterval?: number;
+  name?: string;
+}
+
+/**
+ * Determines if it needs to check for a newer CLI version and returns the last
+ * detected version. The version could be stale, but still newer than the
+ * current version.
+ *
+ * @returns {String|undefined} If a newer version is found, then the lastest
+ * version, otherwise `undefined`.
+ */
+export default function getLatestVersion({
+  cacheDir = XDGAppPaths('com.vercel.cli').cache(),
+  distTag = 'latest',
+  output,
+  pkg,
+  updateCheckInterval = 1000 * 60 * 60 * 24 * 7, // 1 week
+}: GetLatestVersionOptions): string | undefined {
+  if (
+    !pkg ||
+    typeof pkg !== 'object' ||
+    !pkg.name ||
+    typeof pkg.name !== 'string'
+  ) {
+    throw new TypeError('Expected package to be an object with a package name');
+  }
+
+  const cacheFile = resolvePath(
+    cacheDir,
+    'package-updates',
+    `${pkg.name}-${distTag}.json`
+  );
+
+  let cache: PackageInfoCache | undefined;
+  try {
+    cache = readJSONSync(cacheFile);
+  } catch (err: any) {
+    // cache does not exist or malformed
+    if (err.code !== 'ENOENT') {
+      output?.debug(`Error reading latest package cache file: ${err}`);
+    }
+  }
+
+  if (!cache || cache.expireAt < Date.now()) {
+    spawnWorker(
+      {
+        cacheFile,
+        distTag,
+        updateCheckInterval,
+        name: pkg.name,
+      },
+      output
+    );
+  }
+
+  if (
+    cache &&
+    !cache.notified &&
+    pkg.version &&
+    semver.lt(pkg.version, cache.version)
+  ) {
+    cache.notified = true;
+    outputJSONSync(cacheFile, cache);
+    return cache.version;
+  }
+}
+
+/**
+ * Spawn the worker, wait for the worker to report it's ready, then signal the
+ * worker to fetch the latest version.
+ */
+function spawnWorker(
+  payload: GetLatestWorkerPayload,
+  output: Output | undefined
+) {
+  // we need to find the update worker script since the location is
+  // different based on production vs tests
```

Comment:
This part is concerning. We should be testing the actual built output

Reply:
It does test the actual build output. When running the test, `__filename` resolves `packages/cli/src/util/get-latest-version/index.ts` and when running from production, `__filename` resolves to `dist/index.js`. Then it will walk up the directory tree until it hits root (shouldn't happen) or finds `dist/get-latest-worker.js`.

I agree this is gross, but I seem to recall there was an issue with attempting to find `get-latest-worker.js` relative to `get-latest-version/index.ts`.